program.go

```diff
@@ -0,0 +1,263 @@
+
+// Program represents the necessary data structures for a simple XDP program that can filter traffic
+// based on the attached rx queue.
+type Program struct {
+	Program *ebpf.Program
+	Queues  *ebpf.Map
+	Sockets *ebpf.Map
+}
+
+// Attach the XDP Program to an interface.
+func (p *Program) Attach(Ifindex int) error {
+	if err := removeProgram(Ifindex); err != nil {
+		return err
+	}
+	return attachProgram(Ifindex, p.Program)
+}
+
+// Detach the XDP Program from an interface.
+func (p *Program) Detach(Ifindex int) error {
+	return removeProgram(Ifindex)
+}
+
+// Register adds the socket file descriptor as the recipient for packets from the given queueID.
+func (p *Program) Register(queueID int, fd int) error {
+	if err := p.Sockets.Put(uint32(queueID), uint32(fd)); err != nil {
+		return fmt.Errorf("failed to update xsksMap: %v", err)
+	}
+
+	if err := p.Queues.Put(uint32(queueID), uint32(1)); err != nil {
+		return fmt.Errorf("failed to update qidconfMap: %v", err)
+	}
+	return nil
+}
+
+// Unregister removes any associated mapping to sockets for the given queueID.
+func (p *Program) Unregister(queueID int) error {
+	if err := p.Queues.Delete(uint32(queueID)); err != nil {
+		return err
+	}
+	if err := p.Sockets.Delete(uint32(queueID)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close closes and frees the resources allocated for the Program.
+func (p *Program) Close() error {
+	allErrors := []error{}
+	if p.Sockets != nil {
+		if err := p.Sockets.Close(); err != nil {
+			allErrors = append(allErrors, fmt.Errorf("failed to close xsksMap: %v", err))
+		}
+		p.Sockets = nil
+	}
+
+	if p.Queues != nil {
+		if err := p.Queues.Close(); err != nil {
+			allErrors = append(allErrors, fmt.Errorf("failed to close qidconfMap: %v", err))
+		}
+		p.Queues = nil
+	}
+
+	if p.Program != nil {
+		if err := p.Program.Close(); err != nil {
+			allErrors = append(allErrors, fmt.Errorf("failed to close XDP program: %v", err))
+		}
+		p.Program = nil
+	}
+
+	if len(allErrors) > 0 {
+		return allErrors[0]
+	}
+	return nil
+}
+
+// NewProgram returns a translation of the default eBPF XDP program found in the
+// xsk_load_xdp_prog() function in <linux>/tools/lib/bpf/xsk.c:
+// https://github.com/torvalds/linux/blob/master/tools/lib/bpf/xsk.c#L259
+func NewProgram(maxSocketEntries, maxQueueEntries int) (*Program, error) {
+	qidconfMap, err := ebpf.NewMap(&ebpf.MapSpec{
+		Name:       "qidconf_map",
+		Type:       ebpf.Array,
+		KeySize:    uint32(unsafe.Sizeof(int32(0))),
+		ValueSize:  uint32(unsafe.Sizeof(int32(0))),
+		MaxEntries: uint32(maxQueueEntries),
+		Flags:      0,
+		InnerMap:   nil,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ebpf.NewMap qidconf_map failed (try increasing RLIMIT_MEMLOCK): %v", err)
+	}
+
+	xsksMap, err := ebpf.NewMap(&ebpf.MapSpec{
+		Name:       "xsks_map",
+		Type:       ebpf.XSKMap,
+		KeySize:    uint32(unsafe.Sizeof(int32(0))),
+		ValueSize:  uint32(unsafe.Sizeof(int32(0))),
+		MaxEntries: uint32(maxSocketEntries),
```

Comment:
The embedded BPF program uses the `xsksMap` like this
```
bpf_redirect_map(&xsks_map, index, 0)
```
`index` is the queue ID, so the queue ID is used as a key into this map, where the mapped value is the XDP socket  descriptor.

So the line
```
MaxEntries: uint32(maxSocketEntries),
```
actually puts a limit on the number and range of _queue IDs_ that we can use, rather than limiting the number of Sockets that can be connected to this Program.

I found this by experimenting with your changes in a [virtual environment with multi-queue enabled](https://github.com/asavie/xdp/wiki/Setting-up-test-environment).

So my suggestion is to set this to `maxQueueEntries` and do away with `maxSocketEntries` entirely (unless there's a reason you want to limit the maximum number of Sockets that can be connected to Program that I missed?)
(Later we can get rid of the `qidconfMap` - I just noticed that in Linux kernel 5.4 they actually minimized the BPF program, [it only looks up the XDP socket map now](https://github.com/torvalds/linux/blob/f6e1ea19649216156576aeafa784e3b4cee45549/tools/lib/bpf/xsk.c#L361-L378), makes sense...)
Let me know what you think.